tools/update_roles.php

```diff
@@ -0,0 +1,744 @@
+ *
+ * Returns all the user's roles based on their permissions
+ *
+ * @param int $userID the UserID
+ *
+ * @return array
+ */
+function getUserRolesByPermission($userID)
+{
+    global $roles;
+    $userPermissions = getUserPermissions($userID);
+    $userRoles       = array();
+
+    foreach ($roles as $role) {
+        $rolePermissions = getRolePermissions($role);
+        if (!array_diff($rolePermissions, $userPermissions)) {
+            array_push($userRoles, $role);
+        }
+    }
+
+    return $userRoles;
+}
+
+/**
+ * GetUserRoles
+ *
+ * Returns all the user's roles
+ *
+ * @param int $userID the UserID
+ *
+ * @return array
+ */
+function getUserRoles($userID)
+{
+    $DB = Database::singleton();
+
+    $userRoles = $DB->pselect(
+        "SELECT upc.permission_category_id
+         FROM users_permission_categories_rel upc
+         LEFT JOIN users u ON u.ID=upc.user_id
+         WHERE u.UserID=:userID",
+        array(
+         'userID' => $userID,
+        )
+    );
+
+    return $userRoles;
+}
+
+/**
+ * UserHasRole
+ *
+ * Checks if a given user has a given role
+ *
+ * @param int    $userID the user ID
+ * @param string $role   the name of the role
+ *
+ * @return boolean
+ */
+function userHasRole($userID, $role)
+{
+    $DB = Database::singleton();
+
+    $roleID = $DB->pselect(
+        "SELECT upc.permission_category_id
+         FROM users_permission_categories_rel upc
+         LEFT JOIN permission_categories pc ON pc.id=upc.permission_category_id
+         WHERE upc.user_id=:userID AND pc.category=:role",
+        array(
+         'userID' => $userID,
+         'role'   => $role,
+        )
+    );
+
+    if (empty($roleID)) {
+        return false;
+    }
+    return true;
+}
+
+/**
+ * UserHasPermission
+ *
+ * Checks if a given user has a given permission
+ *
+ * @param int    $userID     the user ID
+ * @param string $permission the name of the permission
+ *
+ * @return boolean
+ */
+function userHasPermission($userID, $permission)
+{
+    $DB = Database::singleton();
+
+    $permission = $DB->pselect(
+        "SELECT upr.userID
+         FROM user_perm_rel upr
+         LEFT JOIN permissions p USING (permID)
+         WHERE upr.userID=:userID AND p.code=:permission",
+        array(
+         'userID'     => $userID,
+         'permission' => $permission,
+        )
+    );
+
+    if (empty($permission)) {
+        return false;
+    }
+    return true;
+}
+
+/**
+ * RoleHasPermission
+ *
+ * Checks if a given role contains a given permission
+ *
+ * @param string $role       the name of the role
+ * @param string $permission the name of the permission
+ *
+ * @return boolean
+ */
+function roleHasPermission($role, $permission)
+{
+    $rolePermissions = getRolePermissions($role);
+
+    if (in_array($permission, $rolePermissions)) {
+        return true;
+    }
+    return false;
+}
+
+/**
+ * RefreshRoles
+ *
+ * Rebuilds roles based on user permissions
+ *
+ * @return boolean
+ */
+function refreshRoles()
```

Comment:
seems to remove role when not all permissions are there and adds roles if all permissions are checked (see comment on refresh permissions)